README-comments.md

```diff
@@ -0,0 +1,268 @@
+> Ingress is available for standard clusters only and requires at least two worker nodes in the cluster to ensure high availability and that periodic updates are applied.
+
+So we're going to set up nginx to be the ingress service.
+
+```
+kubectl apply -f https://developer.lightbend.com/docs/lightbend-orchestration-kubernetes/latest/files/example-nginx-controller.yaml
+```
+
+This deploys an 'ingress controller' to distribute the traffic to the correct endpoints and a 'default back end' to handle the URLs that aren't mapped. 
+
+Nginx is running as a service within the Kubernetes cluster, and the Container Service has chosen which port to provide for http and https traffic. So we need to look up those ports:
+
+```
+kubectl get services
+```
+
+and look for the port mappings listed against the nginx-ingress service. For example
+
+```
+NAME                         TYPE           CLUSTER-IP       EXTERNAL-IP   PORT(S)                                           AGE
+activityservice              ClusterIP      172.21.173.94    <none>        10000/TCP                                         18m
+chirpservice                 NodePort       172.21.135.141   <none>        10000:30597/TCP,10001:31203/TCP,10002:30775/TCP   18m
+friendservice                ClusterIP      172.21.179.173   <none>        10000/TCP,10001/TCP,10002/TCP                     18m
+front-end                    ClusterIP      172.21.104.133   <none>        10000/TCP                                         18m
+kubernetes                   ClusterIP      172.21.0.1       <none>        443/TCP                                           52m
+nginx-default-backend        ClusterIP      172.21.57.24     <none>        80/TCP                                            2m
+nginx-ingress                LoadBalancer   172.21.39.132    <pending>     80:32532/TCP,443:31139/TCP                        2m
+reactive-sandbox-cassandra   ClusterIP      None             <none>        9042/TCP                                          30m
+```
+
+This indicates the internal port 80 of nginx-ingress is mapped externally to 32532 and port 443 is mapped externally to 31139.
+
+To determine the public IP address use the following command:
+
+```
+bx cs workers mycluster
+```
+
+The URL to the Chirper application will be at the https port on this IP address. Use that address in a browser, and after accepting the security credentials as ok, you should see:
+
+![Chirper home page](images/Chirper.png)
+
+### Making an update
+
+You're now in a position to make a change to the application and build the change and deploy to Kubernetes in IBM Cloud. You can just build one of the modules, for example:
+
+```
+sbt chirp-impl/docker:publish
+```
+
+and the new Docker image will be published to the Container Registry and because the `imagePullPolicy` has been set to `Always`, Kubernetes will update the containers as soon as they are updated in the registry.
```

Comment:
Is this true? I think you have to cause K8s to do a rollout by patching or redeploying the `Deployment` to trigger this.

I think the best thing from an educational standpoint would be to disclose how to increment the version in `build.sbt` and use `rp generate-kubernetes-deployment ... --akka-cluster-join-existing` as the preferred method of performing upgrades. The [Canary Deployments](https://developer.lightbend.com/docs/lightbend-orchestration-kubernetes/latest/kubernetes-deployment.html#canary-deployments) section has the commands to do this.

The biggest benefit to this is that it's teaching good practices (versioning), which allows users to roll back to older versions if necessary.

Reply:
Right, K8s will not be watching image changes in the registry. `Always` means that image will be pulled when the pod starts up.

Reply:
ok, I'll update for this

Reply:
OK, so the issue is that when doing a `kubectl apply` then the apply and the `docker pull` are only done if something has changed in the Kube yaml since the last time. By using the git commit hash as the tag - and of course that meant a change to the Kube yaml - we were able to force Kube to deploy. There's a [tip here](https://github.com/kubernetes/helm/blob/master/docs/charts_tips_and_tricks.md#automatically-roll-deployments-when-configmaps-or-secrets-change) when Helm is involved, used when a reference configmap.yaml is changed - causing a redeploy despite the kube yaml itself not changing.
Anyway ... not that I see the `--ingress-annotation` flag on `rp` I can see that:
```
--ingress-annotation $(uuid)
```
may work - it depends if just changing the kube yaml causes Kube to re-pull the images from the registry despite the image tag not changing when new images were pushed. Otherwise I'll need to go back to maybe using the git commit. I'll try a few things.